### DISCUSSION TITLE
Testing new changes

### DISCUSSION THREAD

**Discussion by MOshima-PIFSC** [PIFSCstockassessments/JABBA#2] [General]
Hi all,
John has been adding some new features to the PIFSC-dev branch already and while he was working on this, we realized that we need a system to test the changes we add before we push them to the repo. This way, if someone downloads the code and tries to run it, they can be sure it works and won't have to spend a lot of time de-bugging things and trying to figure out what changed. I have a few suggestions for this to get us started but thought it should be something we can all discuss and agree upon here. 

- First, if you want to make changes to the code make sure you have cloned the repository, not just installed the package (e.g. `install_github(PIFSCstockassessments/JABBA, ref = "PIFSC-dev")`). 
- Then when you want to make changes to the code, go into that JABBA folder on your computer and open the JABBA Rproject (icon of light blue cube with an R inside). 
- Make any edits you want to the scripts and when you are finished you can run `devtools::load_all()` or go to Build ->  Load all in the top menu. This will load the functions with the updates. 
- Then run `devtools::check()` to make sure the code is passing all checks. (Note there are currently 2 errors that are produced, these are already in the code so don't worry about those). 

I would also recommend testing the function out in a separate script (don't put this in the R folder as John and I learned from experience) with an existing model or an example model.  Once you are pretty sure everything is working as expected, you can push the changes to the PIFSC-dev branch (make sure you are always working from this branch). If you want to have someone check over your code, you can request someone to review it by tagging them in a comment below the files changed in the commit. 

Another thing we can think about is having a generic example dataset that we can use to test the code. This is pretty standard in a lot of R packages to have unit tests that ensure code works as expected. I see that Henning has created some tests already (in the tests folder) that might work for us as well but I haven't looked at them closely to see what they actually test. 

Anyways, just thought I would start a discussion thread about this so we can think about it and figure out a good way to go about this.

**Comment by tobyematthews** [comment]
To be honest, I don't think I have enough experience with multiple people editing the same code base to suggest any changes to this protocol, but I'll follow the suggestions and am watching this discussion!

**Comment by MOshima-PIFSC** [comment]
I spoke with Felipe and Eric this afternoon and I think we have come up with a suggested protocol for now. We can try it out and see if it is too complicated or confusing and adjust if we need to but I think this will make sure no major errors get introduced by accident. 
We have now made the PIFSC-dev branch protected so if you try to make any changes to it, you need to do so via a pull request. To do this, create a new branch and name it FEAT-whatever-feature-you-are-working-on (e.g. FEAT-overfishing-indicator). Do all of your work in this branch. Here you can make multiple commits and keep working on it until you are finished and no one will be affected. Once you have run the basic checks (devtools::check()) and tests (see below for more info), then submit a pull request to merge your FEAT branch into PIFSC-dev branch. Once you've submitted the pull request (PR), I will make sure there aren't any merge conflicts and resolve them if there are. Once everything looks good I'll merge the branch into the PIFSC-dev branch and the change will be added for everyone to use. 
I am also going to take some time tomorrow to set up a very simple example data set and some automated tests that will compare the output with previous output to make sure you are getting the same results. I'll set it up so that the tests run automatically when you push a commit, so no need to run them on your computer. 
If anyone has any other suggestions or ideas to add feel free to share here.

**Reply by jsyslo** [reply]
Thanks, Meg.  Seems straightforward enough. No suggestions from me right now.
